emily/lambda/src/operations/chainstate.rs

```diff
@@ -2,50 +2,148 @@
+    common::package_response(chainstate_entry, 201)
 }
 
 /// Handles retrieval of the current chainstate
-pub fn handle_get_chainstate(
-    _path_parameters:  HashMap<String, String>
+pub async fn handle_get_chainstate(
+    _path_parameters: HashMap<String, String>,
+    context: &LambdaContext,
 ) -> Result<common::SimpleApiResponse, errors::EmilyApiError> {
-    // path_parameters.keys()
-    Ok(models::GetChainstateResponseContent {
-        block_hash: "Tortoise".to_string(),
-        block_height: 42.0,
-    })
-    .and_then(|response| {
-        common::package_response(response, 200)
-    })
+    // // path_parameters.keys()
+    let height: &String = _path_parameters.get("height").unwrap();
+    let query_output = context.dynamodb_client
+        .query()
+        .table_name(context.settings.chainstate_table_name.clone())
+        .key_condition_expression("#h = :height")
+        .expression_attribute_names("#h", "BlockHeight")
+        .expression_attribute_values(":height", AttributeValue::N(height.clone()))
+        .send()
+        .await
+        // TODO:
+        // Make error provide less internal info.
+        .map_err(|e|
+            EmilyApiError::InternalService(
+                format!("Error occurred for table {:?} {:?}", context.settings.chainstate_table_name.clone(), e)
+            )
+        )?;
+
+    // Change behavior depending on how many entries were found.
+    match query_output.items().len() {
+
+        // No blocks found.
+        0 => Err(EmilyApiError::NotFound(
+            format!("No bock hash found for height {}", height)
+        )),
+
+        // A single block found; happy path.
+        1 => {
+            // Extract.
+            let item = query_output.items().first().unwrap();
+
+            // TODO:
+            // extract fields more safely.
+            let block_hash = item.get("BlockHash").unwrap().as_s().unwrap();
+            let block_height = item.get("BlockHeight").unwrap().as_n().unwrap();
+
+            // Package.
+            common::package_response(models::GetChainstateResponseContent {
+                block_hash: block_hash.clone(),
+                block_height: block_height.parse::<f64>().unwrap(),
+            }, 200)
+        },
+
+        // Multiple conflicting blocks found.
+        _ => {
+            let block_hashes = query_output.items()
+                .iter()
+                .map(|item| {
+                    item.get("BlockHash").unwrap().as_s().unwrap().clone()
+                })
+                .collect::<Vec<String>>();
+
+            Err(EmilyApiError::InternalService(
+                format!(
+                    "Multiple block hashes for height {}: [{}]",
+                    height,
+                    block_hashes.join(", "),
+                ).to_string()
+            ))
+        }
+    }
 }
 
 /// Handles the update of an existing chainstate
-pub fn handle_update_chainstate(
-    body:  Option<String>
+pub async fn handle_update_chainstate(
```

Comment:
This function seems identical to the create handler. Why not just remove it and reuse the other one? Also, why have a POST and a PUT endpoint if they do the same thing?

Reply:
The PUT one is specifically to update an existing chain link, so there will be more logic to it, the POST is explicitly an attempt to add a link that isn't already there.

Reply:
> The PUT one is specifically to update an existing chain link, so there will be more logic to it, the POST is explicitly an attempt to add a link that isn't already there.

Oh right of course, that makes sense.

Thinking about this some more, I think we can remove the PUT call with a slightly different table: (1) add a timestamp for when the row was inserted; (2) Change the sort key to the timestamp.

Then everything can be a PutItem, including data when there is a fork. And for GetItem requests, only take the row with the most recent timestamp at the requested block height.

Reply:
Part of this is a permissions issue; we want people to be able to post a deposit but only the signers can update an existing deposit. We can use the same rust function call but it needs to be a different endpoint.

Reply:
I had thought that we didn't have the permission issue for PUT vs POST endpoints for `chainstate` updates since both were private.